includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -150,6 +151,89 @@ private static function register_callbacks() {
 		do_action( 'amp_extract_image_dimensions_batch_callbacks_registered' );
 	}
 
+	/**
+	 * Extract dimensions from filename if dimension exists or from file system.
+	 *
+	 * @param array $dimensions Image urls mapped to dimensions.
+	 *
+	 * @return array Dimensions mapped to image urls, or false if they could not be retrieved
+	 */
+	public static function extract_by_filename_or_filesystem( $dimensions ) {
+
+		if ( empty( $dimensions ) || ! is_array( $dimensions ) ) {
+			return [];
+		}
+
+		$upload_dir      = wp_get_upload_dir();
+		$base_upload_uri = trim( $upload_dir['baseurl'] );
+
+		foreach ( $dimensions as $url => $value ) {
+
+			// Check whether some other callback attached to the filter already provided dimensions for this image.
+			if ( ! empty( $value ) && is_array( $value ) ) {
+				continue;
+			}
+
+			// Get image dimension from file name.
+			$basename                   = basename( $url );
+			$filename_without_extension = explode( '.', $basename );
+			$extension                  = array_pop( $filename_without_extension );
+			$filename_without_extension = implode( '.', $filename_without_extension );
+
+			$regex = '/-(?<width>\d+)x(?<height>\d+)(?:\.' . $extension . ')$/m';
+			preg_match( $regex, $basename, $possible_dimension );
+
+			if ( ! empty( $possible_dimension['width'] ) && ! empty( $possible_dimension['height'] ) ) {
```

Comment:
I think this can be simplified:

```suggestion
			// Get image dimension from file name.
			$path = wp_parse_url( $url, PHP_URL_PATH );
			if ( empty( $path ) ) {
				continue;
			}
			
			if ( preg_match( '/-(?<width>\d+)x(?<height>\d+)\.\w+$/', $path, $possible_dimension ) ) {
```